modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/maintenance/RebuildIndexAction.java

```diff
@@ -0,0 +1,323 @@
+
+        schemaManager.createIndex(
+            targetTable,
+            targetTable.rowDescriptor().tableDescriptor(),
+            targetTable.getSchema().getName(),
+            newIndex,
+            true,
+            visitor
+        );
+
+        // This future must be already finished by the schema index cache visitor above
+        assert createIdxFut.isDone();
+    }
+
+    /**
+     * Destroys old index.
+     *
+     * @param index Index.
+     * @param table Table.
+     * @throws IgniteCheckedException If failed to destroy index.
+     */
+    private void destroyOldIndex(H2TreeIndex index, GridH2Table table) throws IgniteCheckedException {
+        index.destroy0(true, true);
+
+        Session session = table.getDatabase().getSystemSession();
+
+        table.removeIndex(session, index);
+    }
+
+    /**
+     * Finds index for the cache by the name.
+     *
+     * @param cacheName Cache name.
+     * @param idxName Index name.
+     * @param schemaManager Schema manager.
+     * @return Index or {@code null} if index was not found.
+     */
+    @Nullable private H2TreeIndex findIndex(String cacheName, String idxName, SchemaManager schemaManager) {
+        H2TreeIndex targetIndex = null;
+
+        for (H2TableDescriptor tblDesc : schemaManager.tablesForCache(cacheName)) {
+            GridH2Table tbl = tblDesc.table();
+
+            assert tbl != null;
+
+            Index index = tbl.getIndex(idxName);
+
+            if (index != null) {
+                assert index instanceof H2TreeIndex;
+
+                targetIndex = (H2TreeIndex) index;
+                break;
+            }
+        }
+
+        return targetIndex;
+    }
+
+    /**
+     * Prepares system for the rebuild.
+     *
+     * @param database Database manager.
+     * @param manager Checkpoint manager.
+     * @param storage Index build status storage.
+     * @throws IgniteCheckedException If failed.
+     */
+    private void prepareForRebuild(
+        GridCacheDatabaseSharedManager database,
+        CheckpointManager manager,
+        IndexBuildStatusStorage storage
+    ) throws IgniteCheckedException {
+        // Enable WAL
+        database.resumeWalLogging();
+
+        // Enable checkpointer
+        database.onStateRestored(null);
+
+        // Wait for checkpoint
+        manager.forceCheckpoint("beforeIndexRebuild", null).futureFor(FINISHED).get();
+
+        // Flush WAL
+        database.preserveWalTailPointer();
+
+        // IndexBuildStatusStorage listens for checkpoint to update the status of the rebuild in the metastorage.
+        // We need to set up the listener manually here, because it's maintenance mode.
+        manager.addCheckpointListener(storage, null);
+    }
+
+    /**
+     * Cleans up after the rebuild.
+     *
+     * @param manager Checkpoint manager.
+     * @param storage Index build status storage.
+     */
+    private void cleanUpAfterRebuild(CheckpointManager manager, IndexBuildStatusStorage storage) {
```

Comment:
Maybe at the beginning it is necessary to perform a checkpoint, and then delete the **storage**?